applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_truss_element_3D2N.cpp

```diff
@@ -0,0 +1,332 @@
+                        Vector& rOutput,
+                        const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    if(rVariable == STRESS_ON_GP)
+    {
+        TracedStressType traced_stress_type = static_cast<TracedStressType>(this->GetValue(TRACED_STRESS_TYPE));
+
+        const SizeType  GP_num = (mpPrimalElement->GetGeometry().IntegrationPoints()).size();
+        if (rOutput.size() != GP_num)
+            rOutput.resize(GP_num, false);
+
+        switch (traced_stress_type)
+        {
+            case TracedStressType::FX:
+            {
+                std::vector< array_1d<double, 3 > > force_vector;
+                mpPrimalElement->GetValueOnIntegrationPoints(FORCE, force_vector, rCurrentProcessInfo);
+                for(IndexType i = 0; i < GP_num ; ++i)
+                    rOutput(i) = force_vector[i][0];
+                break;
+            }
+            case TracedStressType::PK2_TRUSS:
+            {
+                std::vector<Vector> stress_vector;
+                mpPrimalElement->GetValueOnIntegrationPoints(PK2_STRESS_VECTOR, stress_vector, rCurrentProcessInfo);
+                for(IndexType i = 0; i < GP_num ; ++i)
+                    rOutput(i) = stress_vector[i][0];
+                break;
+            }
+            default:
+                KRATOS_ERROR << "Invalid stress type! Stress type not supported for this element!" << std::endl;
+        }
+    }
+    else
+    {
+        rOutput.resize(1);
+        rOutput.clear();
+    }
+
+    KRATOS_CATCH("")
+}
+
+void AdjointFiniteDifferenceTrussElement::CalculateStressDisplacementDerivative(const Variable<Vector>& rStressVariable,
+                                    Matrix& rOutput, const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    if(rStressVariable == STRESS_ON_GP)
+    {
+        const SizeType num_nodes = mpPrimalElement->GetGeometry().PointsNumber();
+        const SizeType dimension = mpPrimalElement->GetGeometry().WorkingSpaceDimension();
+        const SizeType num_dofs = num_nodes * dimension;
+        const SizeType num_GP = (mpPrimalElement->GetGeometry().IntegrationPoints()).size();
+        if ( (rOutput.size1() != num_dofs) || (rOutput.size2() != num_GP ) )
+            rOutput.resize(num_dofs, num_GP);
+
+        double derivative_pre_factor;
+        this->GetDerivativePreFactor(derivative_pre_factor, rCurrentProcessInfo);
+
+        Vector length_derivative_vector;
+        this->CalculateCurrentLengthDisplacementDerivative(length_derivative_vector);
+
+        for(IndexType i = 0; i < num_dofs; ++i)
+        {
+             for(IndexType j = 0; j < num_GP; ++j)
+                rOutput(i, j) = length_derivative_vector[i] * derivative_pre_factor;
+        }
+    }
+    else
+        KRATOS_ERROR << "Stress displacement derivative only available for Gauss-points quantities!" << std::endl;
+
+    KRATOS_CATCH("")
+}
+
+int AdjointFiniteDifferenceTrussElement::Check(const ProcessInfo& rCurrentProcessInfo)
```

Comment:
this function should always call the Check-Fct of its parent

Reply:
@armingeiser then we should also call the Check-Fct of the element-base within the ```AdjointFiniteDifferenceBaseElement```?

Reply:
hmm, looks so. I would do it in the next PR where you also deal with the `DomainSize` in the base element.

Reply:
I agree. We should do this within the next upcoming PR.